File: proposed-projects/round-trip-experiment/reverse/20260909T190508Z_7c1544c0/20260910T103012Z_7e3d5183/ported_implementation/src/rules-builder/parse-name.ts
import { GrammarParseError } from '../utils/errors/grammar-parse-error.js';
import { codePoints } from '../utils/js.js';

export const PARSE_NAME_ERROR = 'Failed to find a valid name';

export const GET_INVALID_CHAR_ERROR = (char: string): string =>
  `Invalid character "${char}" when parsing name, ` +
  'only lowercase letters and hyphens are allowed.';

const VALID_CHAR = /[a-zA-Z-]/;
const INVALID_NEXT_CHAR = /[_0-9]/;

const isValidChar = (char: string): boolean => VALID_CHAR.test(char);

const isInvalidNextChar = (char: string): boolean => INVALID_NEXT_CHAR.test(char);

export const parseName = (grammar: string, pos: number): string => {
  const chars = codePoints(grammar);
  let name = '';
  while (pos < chars.length && isValidChar(chars[pos])) {
    name += chars[pos];
    pos += 1;
  }
  if (!name) {
    throw new GrammarParseError(grammar, pos, PARSE_NAME_ERROR);
  }
  if (pos < chars.length && isInvalidNextChar(chars[pos])) {
    throw new GrammarParseError(
      grammar,
      pos,
      GET_INVALID_CHAR_ERROR(chars[pos])
    );
  }
  return name;
};

File: packages/gbnf-experiment/data/20260909T173101Z_8e3da990/ported_implementation/rules_builder/parse_name.py
import re

from ..utils.errors.grammar_parse_error import GrammarParseError

PARSE_NAME_ERROR = 'Failed to find a valid name'


def GET_INVALID_CHAR_ERROR(char: str) -> str:
    return f'Invalid character "{char}" when parsing name, only lowercase letters and hyphens are allowed.'


_VALID_CHAR = re.compile(r'[a-zA-Z-]')
_INVALID_NEXT_CHAR = re.compile(r'[_0-9]')


def _is_valid_char(char: str) -> bool:
    return _VALID_CHAR.search(char) is not None


def _is_invalid_next_char(char: str) -> bool:
    return _INVALID_NEXT_CHAR.search(char) is not None


def parse_name(grammar: str, pos: int) -> str:
    name = ''
    while pos < len(grammar) and _is_valid_char(grammar[pos]):
        name += grammar[pos]
        pos += 1
    if not name:
        raise GrammarParseError(grammar, pos, PARSE_NAME_ERROR)
    if pos < len(grammar) and _is_invalid_next_char(grammar[pos]):
        raise GrammarParseError(grammar, pos, GET_INVALID_CHAR_ERROR(grammar[pos]))
    return name


parseName = parse_name
